umbraco/Umbraco-CMS
# 7383: Fixes urlAliases content finder failing on multilingual websites

### Prerequisites


This fixes #7383 

## Quick description of the issue
If you have a multilingual site and have one-level path domains/hostnames set on the root node (e.g. **/en** & **/cy**)
![image](https://user-images.githubusercontent.com/17008780/109961102-70fc4980-7ce1-11eb-8163-aeba18a53941.png)
And you have a urlAlias on any of your content nodes:
![image](https://user-images.githubusercontent.com/17008780/109961140-7fe2fc00-7ce1-11eb-99a6-a13f8227d080.png)
![image](https://user-images.githubusercontent.com/17008780/109961149-82ddec80-7ce1-11eb-9a68-dc3f405f82df.png)
Then navigating to **"/en/test1"** or **"/en/test-2"**, as shown in the info tab, won't work; as the _ContentFinderByUrlAlias.cs_ is incorrectly including the **"/en"** portion of the URL in its comparison. As you can see here:
![image](https://user-images.githubusercontent.com/17008780/109961394-d4867700-7ce1-11eb-86a3-f86a1c454202.png)

Interestingly, the unit tests were completely missing this issue, as the test cases were actually including the **/en** in their urlAliases! Because of this, I have completely re-written the unit tests for the ContentFinderByUrlAlias to hopefully provide better test coverage.

## Description of the fix
I have added an additional check to the _ContentFinderByUrlAlias_ comparison; It now retrieves the absolute URL of the root node, joins it with the _urlAlias_ of the node being checked, and then compares it with the absolute URL of the current request.

So instead of this comparison:
![FindContentByAlias_old_comparison](https://user-images.githubusercontent.com/17008780/109967805-c8061c80-7ce9-11eb-9de1-f3f1dacf5563.png)

It now does this comparison:
![FindContentByAlias_new_comparison](https://user-images.githubusercontent.com/17008780/109967780-c2103b80-7ce9-11eb-8fc0-ac4980f5288e.png)

![image](https://user-images.githubusercontent.com/17008780/109963415-4c55a100-7ce4-11eb-9e57-2a3c63c207bc.png)

GIF of the urlAlias now working:
![ContentFinderByUrlAlias_working](https://user-images.githubusercontent.com/17008780/109963997-ff25ff00-7ce4-11eb-95e7-e7e8ad229a61.gif)

### Why have I kept the original comparison, instead of only using the new absolute URL comparison?
The original comparison has a slight issue, as it sees this "https://localhost:44331/test1/", instead of https://localhost:44331/en/test1/, as valid URL (you can see this behaviour in the gif on #7383) and fixing that would probably count as a breaking change, so I left that behaviour in.

### Steps to test
- Create a multilingual site
- Set the hostnames of the root node to "/en" and "/cy" (or similar)
- Add some urlAliases to your content nodes
- Attempt to visit the URLs shown in the node's info tab

I have re-written the _ContentFinderByAliasWithDomainsTests_ unit tests to provide more coverage and test a wider range of hostnames. I would also really appreciate it if someone could look over and double-check the new unit tests, just in case I've done something silly or missed an edge case. :)
